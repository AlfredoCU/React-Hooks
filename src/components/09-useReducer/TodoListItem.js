import React, { memo } from "react";
import PropTypes from "prop-types";

const TodoListItem = memo(({ todo, index, handleDelete, handleToggle }) => {
  console.log("Mundo");
  return (
    <li className="list-group-item d-flex justify-content-between align-items-center mb-1">
      <input
        type="checkbox"
        checked={todo.done}
        onChange={() => handleToggle(todo.id)}
      />
      <span>
        {index + 1}. {todo.desc}
      </span>
      <button className="btn btn-danger" onClick={() => handleDelete(todo.id)}>
        Eliminar
      </button>
    </li>
  );
});

TodoListItem.propTypes = {
  todo: PropTypes.object.isRequired,
  index: PropTypes.number.isRequired,
  handleDelete: PropTypes.func.isRequired,
  handleToggle: PropTypes.func.isRequired,
};

export default TodoListItem;
