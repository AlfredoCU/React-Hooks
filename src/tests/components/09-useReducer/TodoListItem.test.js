import React from "react";
import { shallow } from "enzyme";
import "@testing-library/jest-dom";

import { testTodos } from "../../fixtures/testTodos";
import TodoListItem from "../../../components/09-useReducer/TodoListItem";

describe("pruebas al componente TodoListItem", () => {
  // Creamos funciones.
  const handleDelete = jest.fn();
  const handleToggle = jest.fn();

  const wrapper = shallow(
    <TodoListItem
      todo={testTodos[0]}
      index={0}
      handleDelete={handleDelete}
      handleToggle={handleToggle}
    />
  );

  test("Foto del componente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe de llamar a la función handleDelete", () => {
    wrapper.find("button").simulate("click");
    expect(handleDelete).toHaveBeenCalled(); // Si se manda a llamar.
    expect(handleDelete).toHaveBeenCalledTimes(1); // Debe de ser llamado una vez.
    expect(handleDelete).toHaveBeenCalledWith(expect.any(Number)); // El argumento debe de ser un numero.
  });

  test("Debe de llamar a al función handleToggle", () => {
    const value = testTodos[0].id;
    wrapper.find("input").simulate("change", { target: { value } });
    expect(handleToggle).toHaveBeenCalled();
    expect(handleToggle).toHaveBeenCalledTimes(1);
    expect(handleToggle).toHaveBeenCalledWith(value); // También puede ser un testTodo[0].id
  });

  test("Debe de mostrar el texto correctamente", () => {
    const text = wrapper.find("span").text().trim();
    expect(text).toBe(`1. ${testTodos[0].desc}`);
  });
});

// hasClass -> evalua un class de css. true o false.
// prop -> evalua una propiedad.
// props -> evalua todas las propiedades.
