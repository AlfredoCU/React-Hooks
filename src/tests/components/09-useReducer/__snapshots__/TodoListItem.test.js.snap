// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas al componente TodoListItem Foto del componente 1`] = `
<li
  className="list-group-item d-flex justify-content-between align-items-center mb-1"
>
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />
  <span>
    1
    . 
    Aprender React
  </span>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Eliminar
  </button>
</li>
`;
